### PR TITLE
etcdserver, storage, clientv3: watcher ranges

### PR DESCRIPTION
**Pull request by heyitsanthony**
protocol change so watch requests are ranges; server rejects non-prefix ranges

### CODE REVIEW

**Comment by gyuho** [comment]
lgtm. Thanks

**Comment by xiang90** [comment]
@heyitsanthony My concern is that this makes the api incomplete. If we want to do this, we need to somehow support range in watch from the very beginning, no matter how expensive it is.

**Comment by heyitsanthony** [comment]
@xiang90 Fair enough. I want to avoid having to make a new watch request rpc / add more fields the existing watch request rpc down the line when it turns out watch needs ranges.

I think I can hammer out a decent ranged watcher implementation as part of this PR without too much trouble.

**Comment by xiang90** [comment]
> I think I can hammer out a decent ranged watcher implementation as part of this PR without too much trouble.

We do not have to do this in the pr. But this is just something we need to think of. If you want to provide a decent range impl for day 1 of v3. Let's do it.

**Comment by heyitsanthony** [comment]
fairly confident in this interval tree implementation /cc @gyuho @xiang90

File: pkg/adt/interval_tree_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		iv := xy{x, y}
+		if _, ok := ivs[iv]; ok {
+			// don't double insert
+			continue
+		}
+		ivt.Insert(NewInt64Interval(x, y), 123)
+		ivs[iv] = struct{}{}
+	}
+
+	for ab, _ := range ivs {
```
Comment (gyuho) [inline pkg/adt/interval_tree_test.go]: `for ab := range ivs`?

File: pkg/adt/interval_tree_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		for xy, _ := range ivs {
```
Comment (gyuho) [inline pkg/adt/interval_tree_test.go]: `for xy := range ivs`?

File: pkg/adt/interval_tree_test.go (hunk 1 above)
Comment (heyitsanthony) [inline pkg/adt/interval_tree_test.go]: ok

File: pkg/adt/interval_tree_test.go (hunk 2 above)
Comment (heyitsanthony) [inline pkg/adt/interval_tree_test.go]: ok

**Comment by gyuho** [comment]
lgtm. Defer to @xiang90. Thanks.

**Comment by xiang90** [comment]
LGTM once fixed all the issues I mentioned. Thanks!

**Comment by heyitsanthony** [comment]
ok all fixed up; merging on greenlight

**Pull request merged**